Terminal.Gui/View/Adornment/Padding.cs

```diff
@@ -38,4 +38,26 @@ public override ColorScheme ColorScheme
             Parent?.SetNeedsDisplay ();
         }
     }
+
+    /// <inheritdoc/>
+    public override Rectangle FrameToScreen ()
+    {
+        Rectangle ret = base.FrameToScreen ();
+
+        ret.X += Parent != null ? Parent.Margin.Thickness.Left + Parent.Border.Thickness.Left : 0;
+        ret.Y += Parent != null ? Parent.Margin.Thickness.Top + Parent.Border.Thickness.Top : 0;
+
+        return ret;
+    }
+
+    /// <inheritdoc/>
+    public override Thickness GetAdornmentsThickness ()
```

Comment:
This PR are using sub-views on adornments and I need to do this for handling mouse on adornments to get the sub-view which has the mouse over it. If you have other way to do this please show me.

Reply:
> This PR are using sub-views on adornments and I need to do this for handling mouse on adornments to get the sub-view which has the mouse over it. If you have other way to do this please show me.

Somewhere else in this PR I commented that all the complex mouse code in `Application` should be moved into the respective classes (`View` and `Adornment`, specifically). When I wrote `AdornmentHandledMouseEvent` it was a short-term hack. I regret that I did not take the time to engineer how it should have been a method on `Adornment` as I think you thought I thought it was Good (I don't).

We have the opportunity to fix that and do it all correctly.

If I get some time in the next few days, maybe I'll dive into 

- #2814 

Addressing that Issue requires solving the same issues you are working very hard on here. I suspect between the two of us we can converge on a elegant and simple solution. No promises as I've got a lot going on right now...

You are doing really good work on this @BDisp. This is really complex stuff and I appreciate you!

Reply:
I agree it should be moved to their base classes, but firstly I'm trying to deal into the `Application.OnMouseEvent` and when all this is really solid we worry about that, ok?

Reply:
FWIW, I just discovered `Adornment.FrameToScreen` has a bug that is causing `Application.OnMouseEvent` to not work correctly. This may be why you felt you needed to jump through all these hoops!

The correct logic, I think, is:

```cs
   /// <inheritdoc/>
   public override Rectangle FrameToScreen ()
   {
       if (Parent is null)
       {
           return Frame;
       }

       // Adornments are *Children* of a View, not SubViews. Thus View.FrameToScreen will not work.
       // To get the screen-relative coordinates of a Adornment, we need to know who
       // the Parent is
       Rectangle parent = Parent.FrameToScreen ();

       // We now have coordinates relative to our View. If our View's SuperView has
       // a SuperView, keep going...
       return new (new (parent.X + Frame.X, parent.Y + Frame.Y), Frame.Size);
   }
```

Reply:
Also, `AdornmentHandledMouseEvent` should not always return true. This is correct:

```cs
       static bool AdornmentHandledMouseEvent (Adornment? adornment, MouseEventEventArgs args)
       {
           if (adornment?.Thickness.Contains (adornment.FrameToScreen (), args.MouseEvent.X, args.MouseEvent.Y) is not true)
           {
               return false;
           }

           Point boundsPoint = adornment.ScreenToBounds (args.MouseEvent.X, args.MouseEvent.Y);

           var me = new MouseEvent
           {
               X = boundsPoint.X,
               Y = boundsPoint.Y,
               Flags = args.MouseEvent.Flags,
               OfX = boundsPoint.X,
               OfY = boundsPoint.Y,
               View = adornment
           };
           return adornment.OnMouseEvent (me);
       }
```

Reply:
As a result of the above, mouse events for Adornments was never really working correctly, at all. 

Fixing these two issues, breaks `Border` drag and some unit tests. Not sure why yet.

Reply:
That also happened to me when I tried to fix on the `Padding` at the time but break the `Border`. I think the way that `Border` is handling the mouse is a bit hacky maybe.

Reply:
I tried your code but neither the `Border` can be drag and scroll bar isn't showing on `Padding`.

Reply:
> I tried your code but neither the `Border` can be drag and scroll bar isn't showing on `Padding`.

As I noted above, this broke drag. I don't know why. But I am confident my fixes are more correct, so the problem is elsewhere in the drag code.

The `ScrollBar` not showing is probably because you worked around these bugs and didn't realize it.

Reply:
IMPORTANT: I made the fixes above to `v2_develop`, not your PR branch. I didn't have time to see how it would impact your new code.

Reply:
I saw that and I'm having difficult to check what was really was changed.

Reply:
If you could submit a PR to this branch maybe is more easy to find out what is going bad with the Border.

Reply:
Also, the place where we need to put the crux logic is `FindDeepestView`. 

If we do this right, the `view` we get here:

![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/4d98c3f2-fb9d-4426-ace8-8bf9bea09905)

"just works"

Reply:
At first I started coding this in `FindDeepestView`. Then it started to cause problems and become even more complicated. Then I noticed that you already had a condition to capture the Border and so I took advantage of this alternative.
By the way, in which PR did you make the latest changes to `OnMouseEvent`?

Reply:
> By the way, in which PR did you make the latest changes to `OnMouseEvent`?

No PR yet. I just have moments of time to code right now and am split between two machines. So I'm just running experiments on the Adornment fundamentals. 

> At first I started coding this in `FindDeepestView`. Then it started to cause problems and become even more complicated. Then I noticed that you already had a condition to capture the Border and so I took advantage of this alternative. 

I just found another bug. This unit test (should go in `AdornmentsTests`) proves `Adornment.Bounds_get` has a bug:

```cs

    [Fact]
    public void Frames_are_Parent_SuperView_Relative ()
    {
        var view = new View
        {
            X = 1,
            Y = 2,
            Width = 20,
            Height = 31
        };

        var marginThickness = 1;
        view.Margin.Thickness = new Thickness (marginThickness);

        var borderThickness = 2;
        view.Border.Thickness = new Thickness (borderThickness);

        var paddingThickness = 3;
        view.Padding.Thickness = new Thickness (paddingThickness);

        view.BeginInit ();
        view.EndInit ();

        Assert.Equal (new Rectangle (1, 2, 20, 31), view.Frame);
        Assert.Equal (new Rectangle (0, 0, 8, 19), view.Bounds);

        // Margin.Frame is always the same as the view frame
        Assert.Equal (new Rectangle (0, 0, 20, 31), view.Margin.Frame);

        // Border.Frame is View.Frame minus the Margin thickness 
        Assert.Equal (
                      new Rectangle (marginThickness, marginThickness, view.Frame.Width - marginThickness * 2, view.Frame.Height - marginThickness * 2),
                      view.Border.Frame);

        // Padding.Frame is View.Frame minus the Border thickness plus Margin thickness
        Assert.Equal (
                      new Rectangle (
                                     marginThickness + borderThickness,
                                     marginThickness + borderThickness,
                                     view.Frame.Width - (marginThickness + borderThickness) * 2,
                                     view.Frame.Height - (marginThickness + borderThickness) * 2),
                      view.Padding.Frame);
    }

    [Fact]
    public void Bounds_Location_Always_Empty_Size_Correct ()
    {
        var view = new View
        {
            X = 1,
            Y = 2,
            Width = 20,
            Height = 31
        };

        var marginThickness = 1;
        view.Margin.Thickness = new Thickness (marginThickness);

        var borderThickness = 2;
        view.Border.Thickness = new Thickness (borderThickness);

        var paddingThickness = 3;
        view.Padding.Thickness = new Thickness (paddingThickness);

        view.BeginInit ();
        view.EndInit ();

        Assert.Equal (new Rectangle (1, 2, 20, 31), view.Frame);
        Assert.Equal (new Rectangle (0, 0, 8, 19), view.Bounds);

        Assert.Equal (new Rectangle (0, 0, view.Margin.Frame.Width - marginThickness * 2, view.Margin.Frame.Height - marginThickness * 2), view.Margin.Bounds);

        Assert.Equal (new Rectangle (0, 0, view.Border.Frame.Width - borderThickness * 2, view.Border.Frame.Height - borderThickness * 2), view.Border.Bounds);

        Assert.Equal (
                      new Rectangle (
                                     0,
                                     0,
                                     view.Padding.Frame.Width - (marginThickness + borderThickness) * 2,
                                     view.Padding.Frame.Height - (marginThickness + borderThickness) * 2),
                      view.Padding.Bounds);
    }
```

I suspect this is causing you headaches, in addition to the bugs above. The fix:

```cs
    /// <summary>Gets the rectangle that describes the inner area of the Adornment. The Location is always (0,0).</summary>
    public override Rectangle Bounds
    {
        get => new Rectangle (Point.Empty, Thickness?.GetInside (new (Point.Empty, Frame.Size)).Size ?? Frame.Size);
        // QUESTION: So why even have a setter then?
        set => throw new InvalidOperationException ("It makes no sense to set Bounds of a Thickness.");
    }
```

Before, the `get` was `get => Thickness?.GetInside (new (Point.Empty, Frame.Size)) ?? new Rectangle (Point.Empty, Frame.Size);` which returned a non-zero `Bounds.Location` for `Border` and `Padding`.

Note, I have not yet determined what else breaks when this is fixed. But the logic in `v2_develop` here is clearly broken.

If you have time, I suggest you open a PR to fix these `Adornments` bugs in `v2_develop` before continuing on this PR. 

If you don't have time, I can probably get to it in a day or so. I fear you'll be chasing your tail until then.

Reply:
> At first I started coding this in `FindDeepestView`. ...

If I had time, what I'd be doing to fix this (in `v2_develop`, not your PR) is:

- There are no dedicated unit tests for `FindDeepestView` right now. We need some. Write some that test the current `v2_develop` version that prove it works.
  - The current version ignores `Adornments` and returns a view if x/y are anywhere within `view.Frame`. 
  
- Modify `FindDeepestView` to return `Margin`, `Padding`, or `Border` if `x`/`y` match. In `Application.OnMouseEvent`, temporarily ignore these. Eg. like this:

```cs
        var view = View.FindDeepestView (Current, a.MouseEvent.X, a.MouseEvent.Y, out int screenX, out int screenY);

        var view = View.FindDeepestView (Current, a.MouseEvent.X, a.MouseEvent.Y, out int screenX, out int screenY);

        // TODO: Remove this temporary filter:
        if (view is Adornment adornment)
        {
            view = adornment.Parent;
        }
...

- Add `FindDeepestView` unit tests to prove this change works.

- Delete all the `Adornment`-specific code in `Application.OnMouseEvent` and add `MouseEvent` handling code to `Adornment`, `Margin`, `Border`, and `Padding` as appropriate (e.g. `Adornment.MouseClick += Parent.SetFocus()`).

- Don't worry about `Adornment.SubViews` yet. That should happen in either THIS PR or the `Close Box` PR.

Reply:
I had already thought about first correcting this problem in a new PR, before continuing with this one. It really is like it's always chasing my tail :-)